File: week05/2/Prob1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode* now = head;
        ListNode* prev = NULL;

        while(now){
            if(now->val == val){
                if(prev == NULL){
                    // now is head node
                    head = now->next;
                }else{
                    // now is not head node
                    prev->next = now->next;
                    prev = now;
                }
            }else{
                prev = now;
            }
            now = now->next;
        }
        return head;
    }
};

File: week05/2/Prob1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        now = head
        prev = None
        
        while now:
            if now.val == val:
                if prev == None:
                    # prev == None인 경우, now는 head에 위치
                    head = now.next
                else:
                    # prev != None인 경우, now는 중간 노드
                    prev.next = now.next
                    prev = now
            else:
                prev = now
            now = now.next
            
        return head




        